### PR TITLE
Add fetching of provisioned machine's IP

### PR DESCRIPTION
**Pull request by BaptisteGerondeau**
This module adds the support for fetching the provisioned machine's IP from MrP to then feed it to wait_for_connection. Thus, the initial role is executed with "hosts: localhost".

The user used by wait_for_connection must be in the ansible_user option of the host definition, and that is done in the last task of provision.yml and set by machine_user (or root by default).
Please ensure that the preseed you fetch to the machine has the necessary keys, else wait_for_connection will fail. (but you can modify that preseed now so it shouldn't be a problem !)

### CODE REVIEW

**Comment by niedbalski** [comment]
@BaptisteGerondeau I use --connection-type=local or --connection=local and that makes the trick for me, Could you point your use case to exemplify?

**Comment by BaptisteGerondeau** [comment]
My precise use case is getting the ansible to dynamically check that an host, for which we don't have the IP, only the MrP name, is up.

For example, I have a VM that I want provisioned. It's (dynamic) reserved IP in MrP is 10.40.24.1, and when the installer is done I have installed the public key of my local machine on it.
So I just give the role the machine name, it fetches it's reserved IP from MrP and at the end of the playbook it repeatedly tries to connect to the machine.

File: README.md
Comment (chase-qi) [inline README.md:19]: Can you please explain why you need to introduce this new var? Why not just use inventory hostname?

File: library/mr_provisioner_get_ip.py
Comment (chase-qi) [inline library/mr_provisioner_get_ip.py:118]: We are currently maintaining IP address for each host in hosts file. In the past, on thunderx, I remember the reserved IP changed several times, the reason is unknown.

I noticed 'identifier' can be empty, and can be any strings. To benefit from this PR, seems we need to edit each machine's interface to use fixed identifier like 'interface0' across hosts.

Maybe, using 'MAC' as identifier is the most solid choice, we need to main the MAC variable though.

File: library/mr_provisioner_get_ip.py
Comment (chase-qi) [inline library/mr_provisioner_get_ip.py]: I think we can simply return 'lease_ipv4'. When the interface configured as 'dynamic-reserved', 'lease_ipv4' equals to 'configured_ipv4'

File: library/mr_provisioner_get_ip.py
Comment (chase-qi) [inline library/mr_provisioner_get_ip.py:106]: no interface found on the machine

File: tasks/main.yml
Comment (chase-qi) [inline tasks/main.yml]: Why removing it?

File: tasks/provision.yml
Comment (chase-qi) [inline tasks/provision.yml]: Not this topic related? and it is required by the preseed module.

**Review by chase-qi — COMMENTED** [review]
Thanks for the PR! 

From my understanding, I think this feature could be use useful if:
* The lab deployment using mr-provisioner-kea dhcp server plugin.
* We set unified interface identifier across hosts.

Please rebase your changes, and please keep this topic related changes only.

File: README.md
Comment (BaptisteGerondeau) [reply README.md:19]: Inventory hostname entails that you have some kind of DNS/Hostname resolution, in my use case in the inventory file I only have tags and IPs. So the ansible tries to provision a machine named '192.168.1.2' for example, which fails...
Note that this variable defaults to inventory hostname.

File: library/mr_provisioner_get_ip.py
Comment (BaptisteGerondeau) [reply library/mr_provisioner_get_ip.py:118]: It is true that this is the hairy bit, as MrP does not have a unified standard way of naming interfaces...
MAC would be a solution, but then it would require going into the MrP interface, which defeats the purpose of fetching the IP automatically.
To be honest I do not have a good solution to this aside from a consistent standard naming of interfaces, or tailoring your playbook to each specific machine.

File: tasks/main.yml
Comment (BaptisteGerondeau) [reply tasks/main.yml]: Because this precise task needs to be called on the newly created host, which did not exist previously. The initial playbook is called with hosts: localhost, then the 'target' host is dynamically created (not written in the host file), and then the wait_for_connection is called on this 'target' host.
I can put a playbook example on pastebin if need be.

File: tasks/provision.yml
Comment (BaptisteGerondeau) [reply tasks/provision.yml]: My bad.

**Review by danrue — CHANGES_REQUESTED** [review]
My initial impression is that I don't mind more modules in this role, but please do not change the interface to the role itself (i.e. tasks/main.yml). These are backward incompatible changes that impact every existing playbook using this role in non-trivial ways.

However, remember that tasks/main.yml is optional (you can disable it by setting mr_provisioner_do_provision to false), and then you can call each of the modules directly from your own playbook.

Another suggestion is to use role variables that can be enabled to turn on this functionality (and leave the default functionality as it is). These get complicated quickly, so the interface will have to be simple if it's to be accepted.

Lastly, what you have here really looks to me like a dynamic inventory script rather than a role. See http://docs.ansible.com/ansible/devel/user_guide/intro_dynamic_inventory.html. This could be used as a part of your own playbook to determine inventory after running this role.

I really appreciate your contributions and am really happy to see you using this role, but this PR is a -1 from me for the time being.

File: library/mr_provisioner_get_ip.py
Comment (rengolin) [reply library/mr_provisioner_get_ip.py:118]: I have seen this "multiple changes in IP" in our lab, too, and it was gone after I changed the lease time to something more than 5 minutes. The problem is, when you have a lot of machines asking for IPs, Kea gets lost if the lease is too short, because some boots can take more than 5 minutes.

However, if you *always* ask the IP of the machine via MrP, then you *always* get the IP that Kea has given it, no matter how much it changes.

**Comment by rengolin** [comment]
@danrue, I agree backwards compatibility is an issue and we should re-implement it in a more flexible way. However, I don't think we should have a "main" way of doing it which is what one group does, and "other" ways of doing it for other groups.

There are a number of hacks that we had to implement to work around MrP's shortcomings, and I'm sure you guys had to create more than we did. However, as time evolves, I think all groups that use it should converge into a small set of methods.

Hard-coding machine names in /etc/hosts is ugly and honestly, I don't think we should encourage it. However, using DNSmasq could give IPs as well as name resolution, which would work more or less like a hosts file.

I cannot in good conscience recommend Baptiste to implement in the exact hacky way it's there already because it doesn't make sense, so I'd like us all to revisit this and try to move towards the best implementation possible. MrP gives us the IP of the machine and a sensible naming of interfaces gives us the rest, so I think we should try to use that.

Also, I don't want us to have a separate repo to do exactly the same thing just because we implemented in a time where MrP had a bit more features.

**Comment by danrue** [comment]
@rengolin - I agree that this PR could be acceptable if it were more flexible. I tried to suggest some approaches that would satisfy that.

A few points of clarification. I don't recommend /etc/hosts either. When I spoke about hosts, I was referring to ansible's notion of a hosts file. Ansible can do inventory a few ways - a static host file (version controlled along side your playbook), or dynamically using a script (not a role). It does not, however, have good support or even any examples that i've seen that support using a role as an inventory source. See http://docs.ansible.com/ansible/latest/user_guide/intro_inventory.html for an overview of how ansible deals with inventory. If there are examples of this usage in the ansible community, please point them out as I'd be interested in reviewing it.

Ansible roles are designed to run against a host that is defined in ansible's inventory (via a hosts file or dynamic inventory script). Using a role to populate the inventory is possible, but unusual. That's the basis for my objection.

This method of running (using localhost and the role as a dynamic inventory source) is possible, and if there is agreement to do it we can do it as a backward incompatible change (bump the major version number of the role). We could also implement it as a role option, but then that adds to cognitive overhead. Since this role does not have local tests, I am concerned about maintenance over time as it grows in complexity. (due to the recent interest in this role, I'm thinking about adding tests).

**Comment by BaptisteGerondeau** [comment]
I am inclined to agree with @danrue concerning the dynamic inventory script : much cleaner and to the point (and Ansible-y).
There is an example of using the add_host module in the [AWS tutorial](http://docs.ansible.com/ansible/latest/scenario_guides/guide_aws.html) to do about the same thing (i.e. provisioning and then setting up stuff on the provisioned machine).

However, my biggest problem with the dynamic inventory script is how to tell it which interface to use.
From ansible-playbook you cannot pass any arguments to that script, and since this might be a different name per machine (and multiple interfaces per machine) assumptions might not cut it.
So this either needs a flag in MrP (e.g. 'Main interface') or a standardized (re)naming of the interfaces in MrP to then make use of a config file (e.g. 'main_interface=eth0').
We could even imagine trying to ping each of the interfaces to see which one is reachable, but that might entitle an SSH tunnel if the machine's network is not accessible directly to the machine (laptop) on which the Ansible is called.

In the meantime I can add a boolean (e.g. mr_provisioner_get_ip) to the 'wait_for_connection' in the interface (i.e. main.yml) to make it backward compatible.
It is not my desire to break things or increase the amount of maintenance needed for this role, but to have a standardized Ansible interface to MrP as @rengolin pointed out.

**Comment by danrue** [comment]
OK I gave this a try today and I ran into a few issues while trying to use it. Sorry for not putting these comments inline, but I thought this might be easier this time.

1. Please use spaces instead of tabs in README to be consistent with the rest of the file.
2. You don't need 'machine_user' variable; this should be done with ansible inventory and group vars instead.
3. You also don't need mr_provisioner_get_ip, actually. You can just make the conditionals based on whether or not mr_provisioner_machine_name is set.
4. It may be easier to put the 'Get IP of provisioned machine' and 'Add provisioned machine to host' to a separate task file, and then include that task file conditionally from main.yml so that you only have to check the condition once. For example, something like this:
```yaml
- include: provision.yml
  when: mr_provisioner_do_provision
  delegate_to: localhost

- include: add_host.yml
  when: mr_provisioner_machine_name != False

- name: Wait for host for 3600 seconds, but only start checking after 60 seconds
  wait_for_connection:
    delay: 60
    timeout: 3600
  when: mr_provisioner_do_provision and not mr_provisioner_machine_name
```
5. Line 59 in provision.yml ('kernel_options: "{{ mr_provisioner_kernel_options|default('') }}"') produces an ansible error because it is a duplicate of line 57.
6. mr_provisioner_interface_name needs to be added to README.
7. add_host is using hard coded inventory group name 'target' which I don't think is very descriptive. What about something like 'mr_provisioner_hosts'?

Note that the reason all the variables are prepended with mr_provisioner is because, sadly, all vars in ansible are global (I guess they did not learn from history), and so generically named things can easily cause conflict or unexpected behavior when a playbook is using multiple roles.

All that is rather minor, but there is one larger issue too. Support for multiple hosts: the ec2 playbook you mentioned gets around it with an 'exact_count' variable that lets you specify how many hosts to spin up. It seems that we may need a fancier data structure to support multiple hosts (and being able to set i.e. mr_provisioner_interface_name for each one). You could do it as a list of dictionaries.. i.e.
```yaml
ansible_provisioner_machines:
  - hostname: j12-m1-c34n1
    interface_name: eth0
  - hostname: j12-m1-c35n1
    interface_name: eth1
```
Then main.yml needs to implement a loop for each of those machines.
 
I implemented some of the above suggestions here as I was working through them, which you may use as reference or as a cherry-pick directly https://github.com/danrue/ansible-role-mr-provisioner/commit/098939e447f75e3c4d0b856583b81555e08879a5. I didn't try to add multi-host support, yet.

Lastly, the following playbook repository may be useful. It's basically an example, working playbook that exercises this role. I implemented dynamic host support in the [ipgetter-integration branch](https://github.com/danrue/ansible-playbook-mr-provisioner-example/tree/ipgetter-integration)

I'm starting to agree with @rengolin with regard to backward compatibility. As complexity grows, it may be better to just make it how we want it and migrate playbooks. There aren't many playbooks, and this role is versioned so it should not break anything until playbook authors choose to do the upgrade.

**Comment by BaptisteGerondeau** [comment]
Thank you so much for your detailed comments and for taking the time to implement changes directly, I looked into them and cherry-picked them directly.

Multiple host provisioning is an issue I agree, but adopting any kind of changes as you mentioned would also break the interface.
As of Ansible 2.2 you can loop over a role with [include_role](http://docs.ansible.com/ansible/latest/modules/include_role_module.html). I wonder if that would not be more appropriate.
Of course the interface could be changed and a new, non-backwards compatible, version of the role could be put out there.

**Comment by danrue** [comment]
You can do the loop for multiple hosts in the playbook or in the role. If you do it in the role, then playbook authors don't have to worry about it.

We're already breaking the interface. Let's break it once rather than potentially having to break it again in the future in order to support multiple hosts.

I suspect changing it to a datastructure as I proposed and adding a loop in the role will not be difficult.

**Comment by BaptisteGerondeau** [comment]
This change doesn't seem non trivial to me.
I guess it would be nice for the purpose of testing multiple different machines provisioned by the same exact kernel/initrd, but then they might need different kernel command lines (early console), preseeds, etc... So you would need many of the roles variables for each machines, which ends up being like another playbook. If one machine fails to be provisioned, the role is also a failure despite other machines being provisioned.
Then there is the question of how to wait for connection, and how to do it in a parallel manner. Looping over the structure wouldn't allow this.

In fact, for hooking up this role to Jenkins I have a python script that templates a typical playbook with the correct values, then runs it.

In the end, Multiple Hosts provisioning is outside of the scope of this Pull Request.

**Comment by danrue** [comment]
OK. I disagree with the justification a bit but I'm fine with this change at this point because I don't want to hold up progress forever. If @chase-qi also signs off we can merge it.

**Comment by chase-qi** [comment]
Gents, thanks for the efforts! I had a try. Provisioning host from hosts file hit the below error:

```
TASK [Linaro.mr-provisioner : Provision machine] **************************************************************************************************************************************************************************************************************************
fatal: [j12-m1-c34n1 -> localhost]: FAILED! => {"changed": false, "debug": {}, "msg": "Error no assigned machine found with name \"False\""}
fatal: [j12-d05-01 -> localhost]: FAILED! => {"changed": false, "debug": {}, "msg": "Error no assigned machine found with name \"False\""}
```

I made the following changes so that it doesn't break me.

```
diff --git a/defaults/main.yml b/defaults/main.yml
index e0155f7..d68e71a 100644
--- a/defaults/main.yml
+++ b/defaults/main.yml
@@ -5,4 +5,5 @@ mr_provisioner_do_provision: True
 
 # By default, do not fetch the ip of the provisioned machine
 # Instead defer it to the host file, as legacy interface does.
-mr_provisioner_machine_name: False
+# Set the variable to 'True' to add host dynamically.
+mr_provisioner_machine_name: "{{ inventory_hostname }}"
diff --git a/tasks/main.yml b/tasks/main.yml
index 658dfb3..f246551 100644
--- a/tasks/main.yml
+++ b/tasks/main.yml
@@ -3,7 +3,7 @@
   delegate_to: localhost
 
 - include: add_host.yml
-  when: mr_provisioner_machine_name != False and mr_provisioner_do_provision
+  when: mr_provisioner_machine_name == True and mr_provisioner_do_provision
   delegate_to: localhost
   
 - name: Wait for host for 3600 seconds, but only start checking after 60 seconds

```

output example, looks sane?

```
TASK [Linaro.mr-provisioner : Verify variables are set] *******************************************************************************************************************************************************************************************************************
skipping: [j12-d05-01] => (item=mr_provisioner_machine_name) 
skipping: [j12-m1-c34n1] => (item=mr_provisioner_machine_name) 

TASK [Linaro.mr-provisioner : Get IP of provisioned machine] **************************************************************************************************************************************************************************************************************
skipping: [j12-m1-c34n1]
skipping: [j12-d05-01]

TASK [Linaro.mr-provisioner : debug] **************************************************************************************************************************************************************************************************************************************
skipping: [j12-d05-01]
skipping: [j12-m1-c34n1]

TASK [Linaro.mr-provisioner : Add provisioned machine as host] ************************************************************************************************************************************************************************************************************
skipping: [j12-d05-01]

TASK [Linaro.mr-provisioner : Wait for host for 3600 seconds, but only start checking after 60 seconds] **
```

File: tasks/main.yml
Comment (BaptisteGerondeau) [inline tasks/main.yml]: We can safely assume that if the machine_name is the same as the inventory_hostname, not going through the ip getting thing should work well.
If we can't, then a new boolean needs to be introduced, as Ansible doesn't seem to be able to keep track of if a variable has been overridden from defaults.

Comment (chase-qi) [reply tasks/main.yml]: Thanks for the fix. Using inventory_hostname here looks good to me. Here is the only missing bit. Once it is fixed, we are good to merge.
 
```
diff --git a/tasks/main.yml b/tasks/main.yml
index d7cd49a..b5ff621 100644
--- a/tasks/main.yml
+++ b/tasks/main.yml
@@ -10,4 +10,4 @@
   wait_for_connection:
     delay: 60
     timeout: 3600
-  when: mr_provisioner_do_provision and mr_provisioner_machine_name != inventory_hostname
+  when: mr_provisioner_do_provision and mr_provisioner_machine_name == inventory_hostname
```

**Pull request merged**
